temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java

```diff
@@ -100,21 +106,162 @@ public <T> T fromPayloads(
+        }
+        break;
+      default:
+        {
+          // Other type of failure info don't have anything to encode
+        }
+    }
+    return encodedFailure.build();
+  }
+
+  @Override
+  public RuntimeException failureToException(Failure failure) {
+    return dataConverter.failureToException(this.decodeFailure(failure));
+  }
+
+  private Failure decodeFailure(Failure encodedFailure) {
+    Builder decodedFailure = encodedFailure.toBuilder();
+    if (encodedFailure.hasCause()) {
+      decodedFailure.setCause(decodeFailure(encodedFailure.getCause()));
+    }
+    if (encodedFailure.hasEncodedAttributes()) {
+      decodedFailure.setEncodedAttributes(decode(encodedFailure.getEncodedAttributes()));
+    }
+    switch (encodedFailure.getFailureInfoCase()) {
+      case APPLICATION_FAILURE_INFO:
+        {
+          ApplicationFailureInfo encodedInfo = encodedFailure.getApplicationFailureInfo();
+          ApplicationFailureInfo.Builder decodedInfo = encodedInfo.toBuilder();
+          if (encodedInfo.hasDetails()) {
+            decodedInfo.setDetails(decode(encodedInfo.getDetails()));
+          }
+          decodedFailure.setApplicationFailureInfo(decodedInfo);
+        }
+        break;
+      case TIMEOUT_FAILURE_INFO:
+        {
+          TimeoutFailureInfo encodedInfo = encodedFailure.getTimeoutFailureInfo();
+          TimeoutFailureInfo.Builder decodedInfo = encodedInfo.toBuilder();
+          if (encodedInfo.hasLastHeartbeatDetails()) {
+            decodedInfo.setLastHeartbeatDetails(decode(encodedInfo.getLastHeartbeatDetails()));
+          }
+          decodedFailure.setTimeoutFailureInfo(decodedInfo);
+        }
+        break;
+      case CANCELED_FAILURE_INFO:
+        {
+          CanceledFailureInfo encodedInfo = encodedFailure.getCanceledFailureInfo();
+          CanceledFailureInfo.Builder decodedInfo = encodedInfo.toBuilder();
+          if (encodedInfo.hasDetails()) {
+            decodedInfo.setDetails(decode(encodedInfo.getDetails()));
+          }
+          decodedFailure.setCanceledFailureInfo(decodedInfo);
+        }
+        break;
+      case RESET_WORKFLOW_FAILURE_INFO:
+        {
+          ResetWorkflowFailureInfo encodedInfo = encodedFailure.getResetWorkflowFailureInfo();
+          ResetWorkflowFailureInfo.Builder decodedInfo = encodedInfo.toBuilder();
+          if (encodedInfo.hasLastHeartbeatDetails()) {
+            decodedInfo.setLastHeartbeatDetails(decode(encodedInfo.getLastHeartbeatDetails()));
+          }
+          decodedFailure.setResetWorkflowFailureInfo(decodedInfo);
+        }
+        break;
+      default:
+        {
+          // Other type of failure info don't have anything to decode
+        }
+    }
+    return decodedFailure.build();
+  }
+
   @Nonnull
   @Override
   public List<Payload> encode(@Nonnull List<Payload> payloads) {
     return codec.encode(payloads);
   }
 
+  private Payload encode(Payload payload) {
```

Comment:
Could you please give these private encode/decode methods more specific names? The amount of overloads drives me a bit crazy on this piece and I don't think it's really needed here, because they are private. I think it will be more readable if it's `encodePayload` and `encodePayloads`